### ISSUE TITLE
Add support to automatically correct/add plural forms entry to po files (GNA #18108)

### ISSUE THREAD

**Issue by wesnoth-bugs** [wesnoth/wesnoth#1259]
**[Original submission](https://gna.org/bugs/index.php?18108)** by ivanovic on 2011-05-07

Please create a script that automatically checks if the correct plural forms entry is defined for po files. This has to be checking if the entry is available in the specified po file and if it is correct. The entry can be somewhere in the header, the exact position is not fixed, it just has to be after this (while still part of the very first block):

msgid ""

msgstr ""

As an example, please have a look at this commit, doing just this (manually) to the german translation files (svn r49303): http://svn.gna.org/viewcvs/wesnoth?view#rev&revision49303

To get a list of plural forms, please have a look at this website:

http://translate.sourceforge.net/wiki/l10n/pluralforms

It would probably make sense to define the list of "pofile name" and associated plural forms entry in the script itself.

(Reproduced on any)
Release: trunk
Priority: 3 - Low
Severity: 2 - Minor

**Comment by wesnoth-bugs** [comment]
**Modified on 2011-05-11**

**esr wrote:**

> I don't understand why you want a script for this.  It looks to me like Plural-Forms only needs to be added once per .po and can simply live in the master afterwards.

**Comment by wesnoth-bugs** [comment]
**Modified on 2011-05-11**

**ivanovic added attachment:** [langcode-pluralforms.txt](http://gna.org/bugs/download.php?file_id=12993)

**ivanovic wrote:**

> Attaching a plaintext definition for all langcodes that are used in mainline wesnoth and the corresponding plural form entries. Basic idea:
> 
> 1) Check if there is a plural forms inside the LANGCODE.po file
> 
> 2) a) If there is no such entry, add the correct one.
> 
> 2) b) If there is an entry, check if it matches the specified ones. If this is not the case, do not change the file but warn about it (there might be reasons for differing!).
> 
> (file #12993)

**Comment by Wedge009** [comment]
Is this related to #1135?

**Comment by stevecotton** [comment]
> Is this related to #1135?

No, there's no connection other than that they're both about pluralised messages.

**Comment by stevecotton** [comment]
@ivanovic we now have the requested tool, assigning to you to ask what you want to do with the results (which are in a comment above).

**Comment by ivanovic** [comment]
And what exactly am I supposed to do with this, @stevecotton ?
The "challenge" is that translation teams are often resetting things (unintentionally)... Sure, we could apply the correct forms once, but what afterwards?

**Comment by stevecotton** [comment]
> Original submission by ivanovic on 2011-05-07

That's why I was asking you what you wanted to do with it. If neither of us knows, let's just close the issue.

**Comment by lynxlynxlynx** [comment]
As an ex-translator I suggest this be closed. It should be the duty of the translator to add a plural form setting if it is needed.

**Comment by soliton-** [comment]
I don't really understand what the confusion is. The issue is that to this day the plural forms specifications in our po files are inconsistent. Even when just considering the 3 mentioned text domains that actually contain plural messages.

It'd be nice if there was a script (or really CI check nowadays) that would make sure the plural forms specifications are consistent.

https://www.gnu.org/software/gettext/manual/html_node/Plural-forms.html lists perhaps all the languages we have translations for and what the plural forms specification should be? It should also be possible to check that the specifications are as they should be for every given language. I cannot imagine that for wesnoth we would need special plural forms. (Even then we can decide once what that special form is for each language and the check can respect that, too.)

**Comment by lynxlynxlynx** [comment]
You can't check if a plural form spec is right without knowing the language. For example, in Slovenian typically two have been used historically, offset by 1, which influences which plural string applies to which plural. Changing the spec blindly would invalidate most if not all plural translations in the same file. 
Just leave it to the translators and stop worrying.

**Comment by soliton-** [comment]
We do know the language? No idea why you assume some kind of blind change.

It's clear to see right now in the current po files that the specifications are not correct. Of course we can just not care. It's not like anyone did care enough so far and implement this...

No idea what the point would be to pretend everything is fine though. Translators are humans like all other contributors. Helping them make fewer mistakes is a good thing.

**Comment by lynxlynxlynx** [comment]
What do you mean "not correct"? There is no (ISO) standard for them AFAIK, so I wouldn't conflate diversity with incorrectness.

I'm all for mistake prevention, but I don't see how changing plural form specs would contribute to that. On translated files that just creates more work with reshuffling, possibly introducing regressions if not properly done programmatically.

**Comment by soliton-** [comment]
As mentioned earlier the specifications are not consistent. One po file has one specification and another a different one or none at all.

For example:
po/wesnoth/grc.po:       "Plural-Forms: nplurals=3; plural=n==1 ? 0 : n==2 ? 1 : 2;\n"
po/wesnoth-lib/grc.po:   "Plural-Forms: nplurals=INTEGER; plural=EXPRESSION;\n"
With po/wesnoth-httt/grc.po containing no plural forms specification.

**Comment by lynxlynxlynx** [comment]
And I'm saying not being consistent is perfectly fine. If anyone ever picks up the ancient Greek translation again, they will add/adjust the directive as otherwise they won't be able to add plurals. Though in this case of untranslated files, there's no harm in copying the definitions from an already used file.

I'm surprised Wesnoth is still not using Pootle or similar to manage i18n. Things like this would not be an issue.

**Comment by soliton-** [comment]
> And I'm saying not being consistent is perfectly fine.

Disagreed.
